File: Code/1299/1299.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
文件名: PersonalLeetCode
作者: Bolun Xu
创建日期: 2025
版本: 1.0
描述: 将每个元素替换为右侧最大元素 —— 暴力求解。
时间复杂度：O(N)
空间复杂度：O(N)
执行用时：44 ms
消耗内存：18.83 mb
"""
from typing import List

# 暴力求解
# class Solution:
#     def replaceElements(self, arr: List[int]) -> List[int]:
#         ans = [-1] * len(arr)
#         for i in range(len(arr) - 1):
#             ans[i] = max(arr[i + 1:])
#
#         return ans

class Solution:
    def replaceElements(self, arr: List[int]) -> List[int]:
        max_value = -1
        ans = [-1] * len(arr)
        for i in range(len(arr) - 1, -1, -1):
            x = arr[i]
            ans[i] = max_value
            max_value = max(max_value, x)
        return ans


if __name__ == '__main__':
    sol = Solution()
    print(sol.replaceElements(arr=[17, 18, 5, 4, 6, 1]))
    print(sol.replaceElements(arr=[400]))

File: Code/1299/1299.cpp
/******************************************************************************
 * 文件名: 1299
 * 作者: BOLUN XU
 * 创建日期: 2025
 * 版本: 1.0
 * 描述: 将每个元素替换为右侧最大元素。
 * 时间复杂度：O(N)
 * 空间复杂度：O(N)
 * 执行用时：5 ms
 * 消耗内存：71.19 mb
 ******************************************************************************/


#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
  public:
    vector<int> replaceElements(vector<int>& arr) {
      int max_value = -1;
      vector<int> ans(arr.size(), -1);
      for (int i = arr.size() - 1; i >= 0; i--) {
        int x = arr[i];
        ans[i] = max_value;
        max_value = max(max_value, x);
      }
      return ans;

    }
};


int main() {
  Solution s;
  vector<int> arr1 = {17, 18, 5, 4, 6, 1};
  vector<int> result1 = s.replaceElements(arr1);
  cout << "[";
  for (size_t i = 0; i < result1.size(); i++) {
    cout << result1[i];
    if (i < result1.size() - 1) {
      cout << ", ";
    }
  }
  cout << "]" << endl;
  return 0;
}